### ISSUE TITLE
Memory leak in the the Blazor server application

### ISSUE THREAD

**Issue by slavapr** [dotnet/aspnetcore#33547]
### Describe the bug
Noticed a memory leak during Refresh/Close page in the application when there is a custom control that implements IDisposable or IAsyncDisposable where we call InvokeAsync
``` C#
 public class MyCustomControl : ComponentBase, IAsyncDisposable
    {
        [Inject] public IJSRuntime JsRuntime { get; set; }


        public async ValueTask DisposeAsync()
        {
            var token = new CancellationToken(false);
            try
            {
                await JsRuntime.InvokeVoidAsync("MyNamespace.destroyComponent", token);
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw;
            }
            
        }
    }
````
    
AND razor page should contain any control with @bind    
In my example, it is input control.
```
@page "/fetchdata"

@using MemoryLeak.Data
@inject WeatherForecastService ForecastService

<h1>Weather forecast</h1>

<p>This component demonstrates fetching data from a service.</p>

@if (forecasts == null)
{
    <p><em>Loading...</em></p>
}
else
{
    <MyCustomControl/>
    <input @bind="MyValue" />

    <table class="table">
        <thead>
            <tr>
                <th>Date</th>
                <th>Temp. (C)</th>
                <th>Temp. (F)</th>
                <th>Summary</th>
            </tr>
        </thead>
        <tbody>
            @foreach (var forecast in forecasts)
            {
                <tr>
                    <td>@forecast.Date.ToShortDateString()</td>
                    <td>@forecast.TemperatureC</td>
                    <td>@forecast.TemperatureF</td>
                    <td>@forecast.Summary</td>
                </tr>
            }
        </tbody>
    </table>
}

@code {

    private string MyValue { get; set; } = "MyValue";

    private WeatherForecast[] forecasts;

    protected override async Task OnInitializedAsync()
    {
        forecasts = await ForecastService.GetForecastAsync(DateTime.Now);
    }
}
````

To reproduce just click several times Refresh and as you can see WeatherForecast objects are not disposed.
![image](https://user-images.githubusercontent.com/4169617/122064769-11422e00-cdfa-11eb-8799-5181439b3d45.png)

These are info from dotMemory after several Refresh:
![image](https://user-images.githubusercontent.com/4169617/122065275-7ac23c80-cdfa-11eb-967a-6dc449bf62ae.png)
![image](https://user-images.githubusercontent.com/4169617/122065371-90376680-cdfa-11eb-9814-4bd09b4e6477.png)
![image](https://user-images.githubusercontent.com/4169617/122065442-a04f4600-cdfa-11eb-9b26-208efbe97eec.png)

### Exceptions (if any)

Exception.
````
JavaScript interop calls cannot be issued at this time. This is because the component is being statically rendered. When prerendering is enabled, JavaScript interop calls can only be performed during the OnAfterRenderAsync lifecycle method.
````

![image](https://user-images.githubusercontent.com/4169617/122045941-f6b28980-cde6-11eb-8796-126e100a5440.png)

### Further technical details
- ASP.NET Core version
5.0
- Include the output of `dotnet --info`
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.4.21255.9
 Commit:    950e4949a7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-preview.4.21255.9\

Host (useful for support):
  Version: 6.0.0-preview.4.21253.7
  Commit:  bfd6048a60

.NET SDKs installed:
  2.1.805 [C:\Program Files\dotnet\sdk]
  2.2.207 [C:\Program Files\dotnet\sdk]
  3.1.101 [C:\Program Files\dotnet\sdk]
  3.1.300 [C:\Program Files\dotnet\sdk]
  3.1.302 [C:\Program Files\dotnet\sdk]
  3.1.401 [C:\Program Files\dotnet\sdk]
  3.1.402 [C:\Program Files\dotnet\sdk]
  3.1.403 [C:\Program Files\dotnet\sdk]
  3.1.404 [C:\Program Files\dotnet\sdk]
  5.0.101 [C:\Program Files\dotnet\sdk]
  5.0.202 [C:\Program Files\dotnet\sdk]
  5.0.300-preview.21258.4 [C:\Program Files\dotnet\sdk]
  5.0.301 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.4.21255.9 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.4.21253.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.4.21253.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-preview.4.21254.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version
VS2019 Professional 

Solution to reproduce:
[MemoryLeak.zip](https://github.com/dotnet/aspnetcore/files/6656141/MemoryLeak.zip)

**Comment by akovac35** [comment]
Is it possible the leak happens because the javascript calls in the dispose method do not return? Please add a console statement to verify.

**Comment by SteveSandersonMS** [comment]
After investigation, I think this is another effect of https://github.com/dotnet/aspnetcore/issues/32808, which is already fixed for 6.0.

In 5.0 and earlier, if you did a JS interop call after the circuit was disposed, it would take 1 minute to time out. During that time, your instances will remain live. It's not a leak since things are cleared up after the 1 minute period (or sooner if you impose your own shorter timeout), but it is confusing and would explain why it looked like there was a problem.

I ran the repro code and was able to see that, on 6.0, the objects are released for GC immediately. If I navigate away to a different page that calls `GC.Collect()` (to simulate having enough memory pressure for GC to run) then no `WeatherForecast` instances remain in the heap according to the VS memory profiling tools.

So, I'll close this as resolved. If it's an issue for you on 5.0, you might want to change your JS interop call during disposal so that it has a shorter timeout. For example:

```cs
var timeout = TimeSpan.FromSeconds(5);
await JsRuntime!.InvokeVoidAsync("MyNamespace.destroyComponent", timeout, token);
```
